### PR TITLE
Add constraint programming sets

### PR DESCRIPTION
**Pull request by odow**
Part of #1805

This adds:
 * Cumulative
 * BinPacking
 * Table
 * Circuit
 * Path

### CODE REVIEW

File: src/Test/test_cpsat.jl [hunk 1]
```diff
@@ -220,3 +220,175 @@ function setup_test(
...
+    config::Config{T},
+) where {T}
+    @requires MOI.supports_constraint(
+        model,
+        MOI.VectorOfVariables,
+        MOI.BinPacking{T},
+    )
+    @requires MOI.supports_add_constrained_variable(model, MOI.Integer)
+    @requires _supports(config, MOI.optimize!)
+    x = [MOI.add_constrained_variable(model, MOI.Integer())[1] for _ in 1:3]
```
Comment (odow) [inline src/Test/test_cpsat.jl:240]:
```suggestion
    x = [MOI.add_constrained_variable(model, MOI.Integer())[1] for _ in 1:2]
```

File: src/sets.jl [hunk 2]
```diff
@@ -1297,6 +1297,172 @@ struct CountGreaterThan <: AbstractVectorSet
     end
 end
 
+"""
+    BinPacking(capacity::T, weights::Vector{T}) where {T}
+
+The set ``\\{x \\in \\mathbb{R}^d\\}`` such that
```
Comment (blegat) [inline src/sets.jl:1303]: `d`, `c` and `w` are undefined here

File: src/sets.jl [hunk 3]
```diff
@@ -1297,6 +1297,172 @@ struct CountGreaterThan <: AbstractVectorSet
...
+end
+
+function Base.:(==)(x::BinPacking{T}, y::BinPacking{T}) where {T}
+    return x.capacity == y.capacity && x.weights == y.weights
+end
+
+"""
+    Cumulative(dimension::Int)
+
+The set ``\\{(s, d, r, b) \\in \\mathbb{R}^{s+d+r+1}\\}`` representing the
```
Comment (blegat) [inline src/sets.jl:1343]:
```suggestion
The set ``\\{(s, d, r, b) \\in \\mathbb{R}^{length(s)+dimension+length(r)+1}\\}`` representing the
```

**Comment by odow** [comment]
I'm going to keep collapsing these PRs so we can unify them a bit more.

**Pull request merged**